### PR TITLE
992 - Add support for save user settings to data grid

### LINKED ISSUE #992: IdsDataGrid: Save user settings
**Enhancement**

Add save user settings feature like the old version:
https://main-enterprise.demo.design.infor.com/components/datagrid/test-save-settings-manual.html
https://main-enterprise.demo.design.infor.com/components/datagrid/test-save-settings-serverside.html
https://main-enterprise.demo.design.infor.com/components/datagrid/test-save-settings.html

But instead of an object for the settings flatten it to individual settings. Save each of these:

```
 saveUserSettings: {
    columns: true,
    rowHeight: true,
    sortOrder: true,
    pagesize: true,
    activePage: true,
    filter: true
},
```

```
saveColumns, saveFilterRow
```

- fire events in case they want to save it
- give methods to save and restore it
- use locale storage or can save manual

### PR DESCRIPTION
**Pull request by deep7102**
**Explain the details for making this change. What existing problem does the pull request solve?**
Added support for save user settings to data grid

**Related github/jira issue (required)**:
Closes #992

**Steps necessary to review your pull request (required)**:
Pull this branch and build/run the demo app
http://localhost:4300/ids-data-grid/save-settings.html
- Change any setting to be different then origin (reorder columns, go to next page etc...)
- After change it should be saved the current settings to local storage
- Refresh or close/open the page-size
- Data grid should restore as previously saved settings

http://localhost:4300/ids-data-grid/save-settings-manual.html
- Change any setting to be different then origin (reorder columns, go to next page etc...)
- After change it should be saved the current settings to local storage
- Refresh or close/open the page, and then click on Restore settings button
- Data grid should restore as previously saved settings

**Included in this Pull Request**:
- [x] An e2e or functional test for the bug or feature.
- [x] A note to the change log.

### CODE REVIEW

**Review by EdwardCoyle — CHANGES_REQUESTED** [review]
@deep7102 This looks good, but I noticed on http://localhost:4300/ids-data-grid/save-settings-manual.html when you refresh the page, the settings don't seem to persist.  On the other page, everything works as expected

**Review by EdwardCoyle — DISMISSED** [review]
I misunderstood the test steps on the second one.  All looks good here 👍

File: src/components/ids-data-grid/demos/save-settings-manual.html [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+      <ids-layout-grid-cell>
+        <ids-button id="btn-restore" type="secondary">
+          <ids-icon slot="icon" icon="restore-user"></ids-icon>
+          <span slot="text">Restore settings</span>
+        </ids-button>
+        <br/><br/>
+        <ids-text>1. Change any setting to be different then origin (reorder columns, go to next page etc...)</ids-text>
+        <ids-text>2. After change it should be saved the current settings to local storage</ids-text>
+        <ids-text>3. Refresh or close/open the page, and then click on <strong>Restore settings</strong> button</ids-text>
+        <ids-text>4. Data grid should restore as previously saved settings</ids-text>
```
Comment (tmcconechy) [inline src/components/ids-data-grid/demos/save-settings-manual.html:24]:
```suggestion
        <ids-text>4. Data grid should restore what was previously modified in saved settings</ids-text>
```

File: src/components/ids-data-grid/demos/save-settings.html [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+    <ids-layout-grid auto="true">
+      <ids-text font-size="12" type="h1">Data Grid (save user settings)</ids-text>
+    </ids-layout-grid>
+    <ids-layout-grid auto="true">
+      <ids-layout-grid-cell>
+        <ids-text>1. Change any setting to be different then origin (reorder columns, go to next page etc...)</ids-text>
+        <ids-text>2. After change it should be saved the current settings to local storage</ids-text>
+        <ids-text>3. Refresh or close/open the page-size</ids-text>
+        <ids-text>4. Data grid should restore as previously saved settings</ids-text>
```
Comment (tmcconechy) [inline src/components/ids-data-grid/demos/save-settings.html:19]:
```suggestion
        <ids-text>4. Data grid should restore the previously saved settings</ids-text>
```

**Review by tmcconechy — CHANGES_REQUESTED** [review]
Looks great just wondering about

a. if the settings should drop the (s) for correct grammer
b. some minor grammar corrections

I like that this is a mixin.

**Comment by deep7102** [comment]
@tmcconechy updated now, to use as single functions.

File: src/components/ids-data-grid/README.md [hunk 3]
```diff
@@ -229,6 +230,13 @@ When used as an attribute in the DOM the settings are kebab case, when used in J
 - `expandableRowTemplate` {string} Should point to the row `template` element for expandable rows.
 - `treeGrid` {boolean} Indicates a tree grid will be used  in the data grid. See the tree grid section for more details.
 - `groupSelectsChildren` {boolean} If a tree grid has multiple selection, setting this will select all children when a parent is selected.
+- `savesActivePage` {boolean} Sets to saves active page to local storage.
```
[the remark this answers is not in the document]
Comment (deep7102) [reply src/components/ids-data-grid/README.md]: updated now

File: src/components/ids-data-grid/ids-data-grid.ts [hunk 4]
```diff
@@ -146,6 +151,7 @@ export default class IdsDataGrid extends Base {
       attributes.SUPPRESS_ROW_DESELECTION,
       attributes.TREE_GRID,
       attributes.VIRTUAL_SCROLL,
+      attributes.UNIQUE_ID,
```
[the remark this answers is not in the document]
Comment (deep7102) [reply src/components/ids-data-grid/ids-data-grid.ts:154]: its just uniqueId can be use also for something else if we need

**Review by tmcconechy — APPROVED** [review]

**Review by EdwardCoyle — APPROVED** [review]

**Pull request merged**